### PR TITLE
Custom workerfarm, BI-Directional IPC

### LINKED ISSUE #29: Distributed builds
E.g. what if a bundler could run on Lambda? This combined with a [distributed cache](https://github.com/parcel-bundler/parcel/issues/28) would mean that even very large production builds would consistently finish very quickly.

### LINKED ISSUE #637: 🐛 `parcel watch` in 1.5.0 stuck on "Building someFile.js..."
I tried updating my project from parcel-bundler 1.4.1 to 1.5.0, but `parcel watch` doesn't finish. I removed the cache folder manually and started the watch command (see exact CLI command below). It gets stuck at some random file, displaying `∞  Building someFile.js...` forever.

The `parcel build` command works btw.

### 🎛 Configuration (.babelrc, package.json, cli command)

.babelrc:
```js
{
  "presets": ["env", "es2015", "react"],
  "plugins": [
    "transform-class-properties",
    "transform-object-rest-spread",
    "transform-decorators-legacy",
    "transform-es3-member-expression-literals",
    "transform-es3-property-literals"
  ],
  "env": {
    "production": {
      "plugins": ["transform-remove-debugger", "transform-remove-console"]
    }
  }
}
```

package.json (without name etc. but all deps & scripts):
```js
{
  "dependencies": {
    "animate.css": "^3.5.2",
    "babel-polyfill": "^6.26.0",
    "flag-icon-css": "^2.8.0",
    "immer": "^0.3.0",
    "lodash": "^4.17.4",
    "material-ui": "^1.0.0-beta.12",
    "material-ui-icons": "^1.0.0-beta.10",
    "object.values": "^1.0.4",
    "pubsub-js": "^1.5.7",
    "rc-collapse": "^1.7.6",
    "react": "16.2.0",
    "react-dom": "16.2.0",
    "react-json-view": "^1.13.0",
    "react-onclickoutside": "^6.4.0",
    "react-redux": "5.0.6",
    "react-treebeard": "^2.0.3",
    "redux": "3.7.2",
    "redux-saga": "^0.15.6",
    "redux-thunk": "^2.2.0",
    "reselect": "^3.0.1",
    "workerize": "^0.1.2"
  },
  "devDependencies": {
    "babel-eslint": "7.2.3",
    "babel-jest": "^22.0.3",
    "babel-plugin-transform-class-properties": "^6.24.1",
    "babel-plugin-transform-decorators-legacy": "^1.3.4",
    "babel-plugin-transform-es3-member-expression-literals": "^6.22.0",
    "babel-plugin-transform-es3-property-literals": "^6.22.0",
    "babel-plugin-transform-object-rest-spread": "^6.26.0",
    "babel-plugin-transform-remove-console": "^6.8.5",
    "babel-plugin-transform-remove-debugger": "^6.8.5",
    "babel-preset-env": "^1.6.1",
    "babel-preset-es2015": "^6.24.1",
    "babel-preset-react": "^6.24.1",
    "cross-env": "^5.0.2",
    "enzyme": "^2.8.2",
    "eslint": "4.1.1",
    "eslint-config-react-app": "^2.0.1",
    "eslint-plugin-flowtype": "2.34.1",
    "eslint-plugin-import": "2.6.0",
    "eslint-plugin-jsx-a11y": "5.1.1",
    "eslint-plugin-react": "^7.5.1",
    "husky": "^0.14.3",
    "identity-obj-proxy": "^3.0.0",
    "jest": "22.1.2",
    "jest-css-modules": "^1.1.0",
    "lint-staged": "^6.0.0",
    "node-sass": "^4.5.3",
    "parcel-bundler": "1.5.0",
    "prettier": "1.10.2",
    "react-test-renderer": "^16.2.0",
    "redux-mock-store": "^1.2.3",
    "sinon": "^4.1.6"
  },
  "scripts": {
    "watch": "cross-env REACT_APP_VERSION=$npm_package_version parcel watch src/index.js --out-dir ./public/static --public-url ./static/",
    "test": "jest",
    "lint": "eslint ./src",
    "pretty": "prettier --write \"src/**/*.{js,css,md,scss}\"",
    "precommit": "lint-staged",
    "prepush": "npm run test --coverage"
  },
  "lint-staged": {
    "src/**/*.{js,css,md,scss}": [
      "prettier --write",
      "git add"
    ],
    "src/**/*.js": "eslint --max-warnings=0"
  },
  "jest": {
    "rootDir": "src",
    "moduleNameMapper": {
      "\\.(css|less|scss)$": "identity-obj-proxy"
    },
    "transform": {
      "^.+\\.js$": "babel-jest",
      "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/fileTransformer.js"
    },
    "setupFiles": [
      "<rootDir>/setupTests.js"
    ]
  }
}
```

Used cli command: `yarn watch` (see script in package.json above)

### 🤔 Expected Behavior

I expect the parcel watch command to finish building.

### 😯 Current Behavior

The command is getting stuck while building.

### 💁 Possible Solution

No idea, sorry.

### 🔦 Context

I will have to stay at version 1.4.1 until the watch command works again in my project, as I heavily depend on the watch command during development.

### 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 1.5.0 |
| Node             | 8.9.3 |
| npm         | 5.6.0 |
| yarn         | 1.3.2 |
| Operating System | Microsoft Windows 10 Enterprise [Version 10.0.15063] |

### LINKED ISSUE #900: Freeze of bundler since 1.5.1
### 🤔 Expected Behavior

The bundler should finish on any system it runs on.

### 😯 Current Behavior

On specific systems, in our case a virtual Windows Server environment, the bundler stops and random files without any errors.

### 💁 Possible Solution

My investigation showed that it is some wierd scenario of memory issues, causing certain workers to become zombies. Therefore, these workers will never stop working and the bundler stops. It's possible to fix this issue by adjusting the "maxConcurrentCallsPerWorker" settings of the worker-farm. In our case, we reduced this value from 10 to 5 to get a stable result.

To become maximum flexible, I would suggest to make this setting available as env variable, just the like worker count, keeping the default value at 10.

### 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | ^1.5.1
| Node             | ^8
| npm/Yarn         | ^1.3
| Operating System | Virtual Windows Server

### LINKED ISSUE #977: Improve error message when typescript is missing
When referencing a .ts file and running parcel like this:
`parcel index.html`
results in this confusing error:
```
Server running at http://localhost:1234 
🚨  /tmp/parc/index.ts: spawn ENOTDIR
    at _errnoException (util.js:1022:11)
    at ChildProcess.spawn (internal/child_process.js:325:11)
    at Object.exports.spawn (child_process.js:493:9)
    at spawn (/home/luca/.npm-global/lib/node_modules/parcel-bundler/node_modules/cross-spawn/index.js:12:24)
    at Promise (/home/luca/.npm-global/lib/node_modules/parcel-bundler/src/utils/installPackage.js:19:17)
    at Promise (<anonymous>)
    at install (/home/luca/.npm-global/lib/node_modules/parcel-bundler/src/utils/installPackage.js:10:10)
    at **<anonymous>**
```

It can be solved by simply installing typescript locally:
`npm i typescript -D`

The error message should be something like "Typescript should be installed l locally" or "Typescript binary is missing".

### LINKED ISSUE #1029: Issues getting parcel to work correctly on Windows
I've been building a small web app with students this past week, and we used Parcel.  It initially worked really well, and was ideally suited to students new to the idea of bundling.  Parcel does a lot right, and it's very impressive in its simplicity.

However, I was forced to switch to webpack after many students struggled to get things to work on Windows.  Their setups differed from one another (e.g., versions, which shell they used, etc) so it's hard to give you an exact environment spec.  However, it was fairly consistent across the students running Windows, where those on Mac and Linux didn't seem to have so much difficulty.  Two main issues stood out:

* In the build, I had [two `npm` scripts](https://github.com/humphd/bridge-troll/blob/b0468b3f8aa590243f6d5af3389e3711b1c8fbf9/package.json#L13-L14) to run `parcel` with and without defining an environment variable.  The variable was only sometimes defined, and often seemed to lose its value on rebuilds of the code; though I saw it happen on first runs too.

* By far the larger issue was that the build would either never complete, or take forever.  The bundle just never got made, and the browser would load forever.

https://github.com/humphd/bridge-troll/ is the repo I was building with them, and `b0468b3f8aa590243f6d5af3389e3711b1c8fbf9` the last commit before I switched to webpack.

Apologies that this bug is vague, bug @devongovett encouraged me to file something.  I wonder how many of your devs are testing regularly on Windows?  I know I'm on Mac, and it worked perfectly for me there.  But using this more widely requires full Windows support.

If there's nothing actionable here, feel free to close this.  I'll check back in on parcel again as you keep improving things.  It's really cool.

### LINKED ISSUE #1058: Autoinstalling dependencies concurrently
🐛 bug report

### 🤔 Expected Behavior

The workers shouldn't all automatically install dependencies at the same time.

### 😯 Current Behavior

They do.

### 🔦 Context

```
$ yarn start
yarn run v1.5.1
$ parcel src/index.html
Server running at http://localhost:1234
🚨  ...../Phaser.io-v3-Boilerplate/node_modules/phaser/src/renderer/webgl/shaders/ForwardDiffuse.frag: Failed to install glslify-bundle.
[1/4] Resolving packages...
[2/4] Fetching packages...
[2/4] Fetching packages...
error An unexpected error occurred: "https://registry.yarnpkg.com/dargs/-/dargs-5.1.0.tgz: ENOENT: no such file or directory, open '/Users/niklas/Library/Caches/Yarn/v1/npm-dargs-5.1.0-ec7ea50c78564cd36c9d5ec18f66329fade27829/.yarn-tarball.tgz'".
info If you think this is a bug, please open a bug report with the information provided in "...../Phaser.io-v3-Boilerplate/node_modules/phaser/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
error An unexpected error occurred: "https://registry.yarnpkg.com/dargs/-/dargs-5.1.0.tgz: ENOENT: no such file or directory, open '/Users/niklas/Library/Caches/Yarn/v1/npm-dargs-5.1.0-ec7ea50c78564cd36c9d5ec18f66329fade27829/.yarn-tarball.tgz'".
info If you think this is a bug, please open a bug report with the information provided in "...../Phaser.io-v3-Boilerplate/node_modules/phaser/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
    at PromiseQueue.install [as process] (...../parcel2/src/utils/installPackage.js:41:11)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:118:7)
```

```sh
$ ps
  PID TTY           TIME CMD
31374 ttys000    0:00.96 -bash
33028 ttys000    0:00.75 node /usr/local/Cellar/yarn/1.5.1_1/libexec/bin/yarn.js start
33029 ttys000    0:10.91 node /usr/local/bin/parcel src/index.html
33057 ttys000    0:11.55 /usr/local/Cellar/node/9.9.0/bin/node ...../parcel2/node_modules/worker-farm/lib/child/index.js /usr/l
33058 ttys000    0:10.80 /usr/local/Cellar/node/9.9.0/bin/node ...../parcel2/node_modules/worker-farm/lib/child/index.js /usr/l
33424 ttys000    0:01.89 node /usr/local/Cellar/yarn/1.5.1_1/libexec/bin/yarn.js add glslify-bundle -D
33426 ttys000    0:01.91 node /usr/local/Cellar/yarn/1.5.1_1/libexec/bin/yarn.js add glslify-bundle -D
32328 ttys001    0:00.74 -bash
```
### 💻 Code Sample

https://github.com/22mahmoud/Phaser.io-v3-Boilerplate (at dfaab01b4f94b4ada8da0892bc02ab527aa9cc17)

### 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | master
| Node             | 9.9.0
| npm/Yarn         | 1.5.2
| Operating System | macOS 10.12.6

### LINKED ISSUE #1159: Prevent parcel from failing within a worker
# 💬 Request For Comment

I've tested parcel inside a server managed by pm2 and converting parcel tests to ava, in both cases this breaks due to these 2 applications also using workers, resulting in `process.send` being a function even within the local worker/master process

I've opened up this issue to discuss an approach to fix this within the new workerfarm.

## Some possible solutions
* Adding an environment variable to the workers (inside our workerfarm `child.js` code) that states it's a parcel worker, we can simple do the check like `(process.send && process.env.workerType === 'parcel')`
* Removing the use of local worker (might not be such a good idea unless we can get the worker startup time to near 0ms)

Would love to hear other approaches to fixing this :)

Related to my comment on #1143 and the workerfarm branch #1105

### PR DESCRIPTION
**Pull request by DeMoorJasper**
* Adds bi-directional ipc
* Removes node-worker-farm, replacing it with a fully customisable promise based workerfarm
* Fixes localRequire being parallel, Closes #1068 Closes #977 Closes #1058
* Improves stability & performance slightly
* Fixes `IPC_CHANNEL_CLOSED` error at the end of tests
* Adds tests to test the workerfarm
* Fixes freezing workers on windows, Closes #900 Closes #1029 Closes #637 Closes #29 Closes #901 
* Closes #1159

### CODE REVIEW

**Comment by DeMoorJasper** [comment]
Locally all tests pass on mac
And on windows the same ones fail that fail in the master branch.
Not sure what's up with CI

**Comment by DeMoorJasper** [comment]
Tests pass locally on windows, mac, ubuntu server (with 500mb of ram and 4 cores).
The tests that fail on travis also fail in master branch.
Everything should work, feel free to test

**Comment by davidnagli** [comment]
So tests pass on Node 9, but when I tried it on Node 6 I got them to fail.

**Comment by DeMoorJasper** [comment]
All tests that pass in master should pass in here now (node 8 & 6) + some that don't pass in master.
Tests have been added to make sure the workerfarm is reliable.

**Comment by devongovett** [comment]
Nice work. I'll take a closer look soon! 🎉

**Comment by DeMoorJasper** [comment]
Pushed a fix for windows freeze bug to this branch.
Would be nice to have a minor 1.7.1 release after this merge as it fixes a lot of bugs and the windows bug blocks any decent parcel usage on windows.

File: src/workerfarm/worker.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+require('v8-compile-cache');
```
Comment (devongovett) [inline src/workerfarm/worker.js]: move this file out of the folder since it is parcel specific

**Comment by reflog** [comment]
really looking forward to a merge of this PR. parcel is hanging on Windows quite often, so this is a godsend!

File: src/workerfarm/WorkerFarm.js [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+  constructor(options, farmOptions = {}) {
+    super();
+    this.options = Object.assign(
+      {
+        maxConcurrentWorkers: WorkerFarm.getNumWorkers(),
+        maxConcurrentCallsPerWorker: 10,
+        forcedKillTime: 100,
+        warmWorkers: true,
+        useLocalWorker: true,
+        workerPath: require.resolve('../worker')
```
Comment (DeMoorJasper) [inline src/workerfarm/WorkerFarm.js]: This will fail in Node 6 as the test environment runs using babel-register and therefore out of src instead of lib

**Comment by reflog** [comment]
Thank you!

**Comment by tinchoz49** [comment]
Hi everyone! how are you?

Is there any breaking change in this PR related with how we can define a custom asset?

I'm using [parcel-plugin-typescript](https://github.com/fathyb/parcel-plugin-typescript) and with the version 1.7.1 of parcel I'm getting this error: https://github.com/fathyb/parcel-plugin-typescript/issues/36

But it seems that everything is ok with the plugin and it was working fine until the version 1.7.0 of parcel.

**Comment by DeMoorJasper** [comment]
@tinchoz49 Yes, this PR has a stricter error handling than the previous implementation, therefore unknown ipc communication results into throwing an error.
This should be fixed if #1235 lands, as plugins can inject functions into the parcel workers

**Pull request merged**